program/tests/harvest_rewards.rs

```diff
@@ -422,27 +420,214 @@ async fn fail_holder_rewards_invalid_data() {
+)]
+#[test_case(
+    System {
+        token_supply: 0,
+        total_rewards: 10_000,
+        pool_excess_lamports: 1_000, // Not enough.
+    },
+    Holder {
+        token_account_balance: 0,
+        last_seen_total_rewards: 5_000, // Last saw half.
+        unharvested_rewards: 2_500, // Not yet harvested.
+    },
+    1_000;
+    "token supply is zero, some unharvested, half last seen, pool underfunded, pool excess"
+)]
+#[test_case(
+    System {
+        token_supply: 10_000,
+        total_rewards: 10_000,
+        pool_excess_lamports: 20_000,
+    },
+    Holder {
+        token_account_balance: 5_000,
+        last_seen_total_rewards: 0,
+        unharvested_rewards: 0,
+    },
+    5_000;
+    "50% of token supply, 0 unharvested, 0 last seen, pool has enough, 50% of rewards"
+)]
+#[test_case(
+    System {
+        token_supply: 10_000,
+        total_rewards: 10_000,
+        pool_excess_lamports: 2_000, // Not enough.
+    },
+    Holder {
+        token_account_balance: 5_000,
+        last_seen_total_rewards: 0,
+        unharvested_rewards: 0,
+    },
+    2_000; // Pool excess.
+    "50% of token supply, 0 unharvested, 0 last seen, pool underfunded, pool excess"
+)]
+#[test_case(
+    System {
+        token_supply: 10_000,
+        total_rewards: 10_000,
+        pool_excess_lamports: 20_000,
+    },
+    Holder {
+        token_account_balance: 5_000,
+        last_seen_total_rewards: 5_000, // Last saw half.
+        unharvested_rewards: 0, // Harvested since last seen.
+    },
+    2_500;
+    "50% of token supply, 0 unharvested, half last seen, pool has enough, half of 50% of rewards"
+)]
+#[test_case(
+    System {
+        token_supply: 10_000,
+        total_rewards: 10_000,
+        pool_excess_lamports: 1_000, // Not enough.
+    },
+    Holder {
+        token_account_balance: 5_000,
+        last_seen_total_rewards: 5_000, // Last saw half.
+        unharvested_rewards: 0, // Harvested since last seen.
+    },
+    1_000;
+    "50% of token supply, 0 unharvested, half last seen, pool underfunded, pool excess"
+)]
+#[test_case(
+    System {
+        token_supply: 10_000,
+        total_rewards: 10_000,
+        pool_excess_lamports: 20_000,
+    },
+    Holder {
+        token_account_balance: 5_000,
+        last_seen_total_rewards: 5_000, // Last saw half.
+        unharvested_rewards: 2_500, // Not yet harvested.
+    },
+    5_000;
+    "50% of token supply, some unharvested, half last seen, pool has enough, 50% of rewards"
+)]
+#[test_case(
+    System {
+        token_supply: 10_000,
+        total_rewards: 10_000,
+        pool_excess_lamports: 1_000, // Not enough.
+    },
+    Holder {
+        token_account_balance: 5_000,
+        last_seen_total_rewards: 5_000, // Last saw half.
+        unharvested_rewards: 2_500, // Not yet harvested.
+    },
+    1_000;
+    "50% of token supply, some unharvested, half last seen, pool underfunded, pool excess"
+)]
+#[tokio::test]
```

Comment:
Can you add a case for `pool_excess_lamports = 0` but also where unharvested rewards and last seen will get updated?

Reply:
Yep!